crypto/ec/ecp_nistp384.c

```diff
@@ -0,0 +1,1937 @@
+    felem_mul_reduce(ftmp, ftmp2, ftmp); /* 2^287 + ... + 2^33 + 2^31 + ... + 2^0 */
+
+    for (i = 0; i < 94; i++)
+        felem_square_reduce(ftmp, ftmp); /* 2^381 + ... + 2^127 + 2^125 + ... + 2^94 */
+    felem_mul_reduce(ftmp, ftmp6, ftmp); /* 2^381 + ... + 2^127 + 2^125 + ... + 2^94 + 2^29 + ... + 2^0 */
+
+    for (i = 0; i < 2; i++)
+        felem_square_reduce(ftmp, ftmp); /* 2^383 + ... + 2^129 + 2^127 + ... + 2^96 + 2^31 + ... + 2^2 */
+    felem_mul_reduce(ftmp, in, ftmp);    /* 2^383 + ... + 2^129 + 2^127 + ... + 2^96 + 2^31 + ... + 2^2 + 2^0 */
+
+    memcpy(out, ftmp, sizeof(felem));
+}
+
+/*
+ * Zero-check: returns a limb with all bits set if |in| == 0 (mod p)
+ * and 0 otherwise. We know that field elements are reduced to
+ * 0 < in < 2p, so we only need to check two cases:
+ * 0 and 2^384 - 2^128 - 2^96 + 2^32 - 1
+ *   in[k] < 2^56, k < 6
+ *   in[6] <= 2^48
+ */
+static limb felem_is_zero(const felem in)
+{
+    limb zero, p384;
+
+    zero = in[0] | in[1] | in[2] | in[3] | in[4] | in[5] | in[6];
+    zero = ((int64_t) (zero) - 1) >> 63;
+    p384 = (in[0] ^ 0x000000ffffffff) | (in[1] ^ 0xffff0000000000)
+         | (in[2] ^ 0xfffffffffeffff) | (in[3] ^ 0xffffffffffffff)
+         | (in[4] ^ 0xffffffffffffff) | (in[5] ^ 0xffffffffffffff)
+         | (in[6] ^ 0xffffffffffff);
+    p384 = ((int64_t) (p384) - 1) >> 63;
+
+    return (zero | p384);
+}
+
+static int felem_is_zero_int(const void *in)
+{
+    return (int)(felem_is_zero(in) & ((limb) 1));
+}
+
+/*-
+ * felem_contract converts |in| to its unique, minimal representation.
+ * Assume we've removed all redundant bits.
+ * On entry:
+ *   in[k] < 2^56, k < 6
+ *   in[6] <= 2^48
+ */
+static void felem_contract(felem out, const felem in)
+{
+    static const int64_t two56 = ((limb) 1) << 56;
+
+    /*
+     * We know for a fact that 0 <= in < 2*p, for p = 2^384 - 2^128 - 2^96 + 2^32 - 1
+     * Perform two successive, idempotent subtractions to reduce if in >= p.
+     */
+
+    int64_t tmp[NLIMBS], a;
+    unsigned int i;
+
+    for (i = 0; i < NLIMBS; i++)
+        tmp[i] = in[i];
+ 
+    /* Case 1: a = 1 iff in >= 2^384 */
+    a = (in[6] >> 48);
+    tmp[0] += a;
+    tmp[0] -= a << 32;
+    tmp[1] += a << 40;
+    tmp[2] += a << 16;
+    tmp[6] &= 0x0000ffffffffffff;
+
+    /*
+     * eliminate negative coefficients: if tmp[0] is negative, tmp[1] must be
+     * non-zero, so we only need one step
+     */
+    a = tmp[0] >> 63;
+    tmp[0] += a & two56;
+    tmp[1] -= a & 1;
+
+    /* Carry 1 -> 2 -> 3 -> 4 -> 5 -> 6 */
+    tmp[2] += tmp[1] >> 56;
+    tmp[1] &= 0x00ffffffffffffff;
+
+    tmp[3] += tmp[2] >> 56;
+    tmp[2] &= 0x00ffffffffffffff;
+
+    tmp[4] += tmp[3] >> 56;
+    tmp[3] &= 0x00ffffffffffffff;
+
+    tmp[5] += tmp[4] >> 56;
+    tmp[4] &= 0x00ffffffffffffff;
+
+    tmp[6] += tmp[5] >> 56; /* tmp[6] < 2^48 */
+    tmp[5] &= 0x00ffffffffffffff;
+
+    /*
+     * Case 2: a = all ones if p <= in < 2^384, 0 otherwise
+     */
+
+    int64_t cond[5];
```

Comment:
Please move this up to the beginning of the function, we don't mix code and declarations.